test/boost/expr_test.cc

```diff
@@ -1419,6 +1419,77 @@ BOOST_AUTO_TEST_CASE(prepare_cast_bind_var_no_receiver) {
     BOOST_REQUIRE(expr::type_of(prepared) == utf8_type);
 }
 
+// Test preparing (text)? with a known text receiver.
+// Corresponds to `text_col = (text)?`
+BOOST_AUTO_TEST_CASE(prepare_cast_bind_var_text_type_hint_and_text_receiver) {
+    schema_ptr table_schema = make_simple_test_schema();
+    auto [db, db_data] = make_data_dictionary_database(table_schema);
+
+    expression cast_expr =
+        cast{.arg = bind_variable{.bind_index = 0, .receiver = nullptr}, .type = cql3_type::raw::from(utf8_type)};
+
+    expression prepared = prepare_expression(cast_expr, db, "test_ks", table_schema.get(), make_receiver(utf8_type));
+
+    // Can't do a direct comparison because we don't have prepared.arg.receiver
+    BOOST_REQUIRE(is<cast>(prepared) && is<bind_variable>(as<cast>(prepared).arg));
+    ::lw_shared_ptr<column_specification> bind_var_receiver = as<bind_variable>(as<cast>(prepared).arg).receiver;
+    BOOST_REQUIRE(bind_var_receiver->type == utf8_type);
+
+    expression expected = cast{.arg = bind_variable{.bind_index = 0, .receiver = bind_var_receiver}, .type = utf8_type};
+    BOOST_REQUIRE_EQUAL(prepared, expected);
+
+    BOOST_REQUIRE(expr::type_of(prepared) == utf8_type);
+}
+
+// Test preparing (int)? with a known int receiver.
+// Corresponds to `int_col = (int)?`
+BOOST_AUTO_TEST_CASE(prepare_cast_bind_var_int_type_hint_and_int_receiver) {
+    schema_ptr table_schema = make_simple_test_schema();
+    auto [db, db_data] = make_data_dictionary_database(table_schema);
+
+    expression cast_expr =
+        cast{.arg = bind_variable{.bind_index = 0, .receiver = nullptr}, .type = cql3_type::raw::from(int32_type)};
+
+    expression prepared = prepare_expression(cast_expr, db, "test_ks", table_schema.get(), make_receiver(int32_type));
+
+    // Can't do a direct comparison because we don't have prepared.arg.receiver
+    BOOST_REQUIRE(is<cast>(prepared) && is<bind_variable>(as<cast>(prepared).arg));
+    ::lw_shared_ptr<column_specification> bind_var_receiver = as<bind_variable>(as<cast>(prepared).arg).receiver;
+    BOOST_REQUIRE(bind_var_receiver->type == int32_type);
+
+    expression expected =
+        cast{.arg = bind_variable{.bind_index = 0, .receiver = bind_var_receiver}, .type = int32_type};
+    BOOST_REQUIRE_EQUAL(prepared, expected);
+
+    BOOST_REQUIRE(expr::type_of(prepared) == int32_type);
+}
+
+// Test preparing (text)? with a known int receiver.
+// Corresponds to `int_col = (text)?`
+BOOST_AUTO_TEST_CASE(prepare_cast_bind_var_text_type_hint_and_int_receiver) {
+    schema_ptr table_schema = make_simple_test_schema();
+    auto [db, db_data] = make_data_dictionary_database(table_schema);
+
+    expression cast_expr =
+        cast{.arg = bind_variable{.bind_index = 0, .receiver = nullptr}, .type = cql3_type::raw::from(utf8_type)};
+
+    BOOST_REQUIRE_THROW(prepare_expression(cast_expr, db, "test_ks", table_schema.get(), make_receiver(int32_type)),
+                        exceptions::invalid_request_exception);
```

Comment:
This throws, as I want but not as I expect from previous discussion. Ah, it's only blob types that we can decay into. Sort of like void* in C.

Reply:
Yes, when casting an int to blob we can be sure that we can just take the 4 bytes and reinterpret them as a blob.
With `int` and `text` it's not that simple - not every `int` is a valid `text`, but also it doesn't make much sense to cast `text` to `int`.
I guess we treat `blobs` as a bit special, since it makes *some* sense to cast anything to `blob`.